x/paychan/handler.go

```diff
@@ -0,0 +1,296 @@
+package paychan
+
+import (
+	"fmt"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/orm"
+	"github.com/iov-one/weave/x"
+	"github.com/iov-one/weave/x/cash"
+)
+
+const (
+	createPaymentChannelCost   int64 = 300
+	transferPaymentChannelCost int64 = 5
+)
+
+func RegisterRoutes(r weave.Registry, auth x.Authenticator, cash cash.Controller) {
+	bucket := NewPaymentChannelBucket()
+	r.Handle(pathCreatePaymentChannelMsg, &createPaymentChannelHandler{auth: auth, bucket: bucket, cash: cash})
+	r.Handle(pathTransferPaymentChannelMsg, &transferPaymentChannelHandler{auth: auth, bucket: bucket, cash: cash})
+	r.Handle(pathClosePaymentChannelMsg, &closePaymentChannelHandler{auth: auth, bucket: bucket, cash: cash})
+}
+
+type createPaymentChannelHandler struct {
+	auth   x.Authenticator
+	bucket PaymentChannelBucket
+	cash   cash.Controller
+}
+
+var _ weave.Handler = (*createPaymentChannelHandler)(nil)
+
+func (h *createPaymentChannelHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (weave.CheckResult, error) {
+	var res weave.CheckResult
+	if _, err := h.validate(ctx, db, tx); err != nil {
+		return res, err
+	}
+
+	res.GasAllocated += createPaymentChannelCost
+	return res, nil
+}
+
+func (h *createPaymentChannelHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*CreatePaymentChannelMsg, error) {
+	rmsg, err := tx.GetMsg()
+	if err != nil {
+		return nil, err
+	}
+	msg, ok := rmsg.(*CreatePaymentChannelMsg)
+	if !ok {
+		return nil, errors.ErrUnknownTxType(rmsg)
+	}
+
+	if err := msg.Validate(); err != nil {
+		return msg, err
+	}
+
+	// Ensure that the timeout is in the future.
+	if height, _ := weave.GetHeight(ctx); msg.Timeout <= height {
+		return msg, ErrInvalidTimeout(msg.Timeout)
+	}
+
+	if !h.auth.HasAddress(ctx, msg.SenderPublicKey.Address()) {
+		return msg, errors.ErrUnauthorized()
+	}
+
+	return msg, nil
+}
+
+func (h *createPaymentChannelHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (weave.DeliverResult, error) {
+	var res weave.DeliverResult
+	msg, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return res, err
+	}
+
+	obj, err := h.bucket.Create(db, &PaymentChannel{
+		SenderPublicKey: msg.SenderPublicKey,
+		Recipient:       msg.Recipient,
+		Total:           msg.Total,
+		Timeout:         msg.Timeout,
+		Memo:            msg.Memo,
+		Transferred: &x.Coin{
+			Whole:      0,
+			Fractional: 0,
+			Ticker:     msg.Total.Ticker,
+		},
+	})
+	if err != nil {
+		return res, err
+	}
+
+	// Move coins from sender account and deposit total amount available on
+	// that channels account.
+	src := msg.SenderPublicKey.Address()
```

Comment:
req: the src address is the senders address. The `SenderPublicKey` is only used to verify the `Payment` messages being signed proper.